File: examples/react/cases/tasks/task_list/task_list.agentic.case.ts
import {
  type AgenticContext,
  type AgenticDiscovery,
  type AgenticExecutionContext,
  type AgenticExample,
  type AgenticMcpContract,
  type AgenticPolicy,
  type AgenticPrompt,
  type AgenticRagContract,
  type AgenticToolContract,
  BaseAgenticCase,
} from "../../../core/agentic.case";
import { type ApiResponse } from "../../../core/api.case";
import {
  AppCaseError,
  toAppCaseError,
} from "../../../core/shared/app_structural_contracts";
import {
  TaskListDomain,
  type TaskListInput,
  type TaskListOutput,
} from "./task_list.domain.case";

type ExpectedCasesMap = {
  tasks?: {
    task_list?: {
      api?: {
        handler(input: TaskListInput): Promise<ApiResponse<TaskListOutput>>;
      };
    };
  };
};

export class TaskListAgentic extends BaseAgenticCase<
  TaskListInput,
  TaskListOutput
> {
  protected domain(): TaskListDomain {
    return new TaskListDomain();
  }

  public discovery(): AgenticDiscovery {
    return {
      name: this.domainCaseName() ?? "task_list",
      description:
        this.domainDescription() ??
        "List persisted task cards for board rendering.",
      category: "tasks",
      tags: ["tasks", "listing", "board"],
      aliases: ["list_board_tasks", "show_board", "inspect_board_state"],
      capabilities: ["task_listing", "board_grounding"],
      intents: ["list tasks", "show the board", "show current tasks"],
    };
  }

  public context(): AgenticExecutionContext {
    return {
      requiresAuth: false,
      dependencies: ["task_list.domain", "task_list.api"],
      preconditions: ["The persisted task store must be readable."],
      constraints: [
        "This capability is read-only.",
        "No filters or pagination are supported in v1.",
      ],
      notes: [
        "Use this capability to ground follow-up decisions before mutating the board.",
      ],
    };
  }

  public prompt(): AgenticPrompt {
    return {
      purpose: "List all persisted tasks so an agent can inspect the board state.",
      whenToUse: [
        "When the user asks to see the board or current tasks.",
        "Before moving a task when the user has not provided an exact task identifier.",
      ],
      whenNotToUse: [
        "When the user wants to create a new task.",
        "When the user already provided a precise task id for a direct move operation.",
      ],
      constraints: [
        "Do not claim support for filters or search in this v1 example.",
      ],
      reasoningHints: [
        "Treat task_list as the canonical grounding step before ambiguous task mutations.",
      ],
      expectedOutcome:
        "A flat array of task objects ordered by createdAt descending.",
    };
  }

  public tool(): AgenticToolContract<TaskListInput, TaskListOutput> {
    const inputSchema = this.domainInputSchema();
    const outputSchema = this.domainOutputSchema();

    if (!inputSchema || !outputSchema) {
      throw new Error("task_list.agentic requires domain schemas");
    }

    return {
      name: "task_list",
      description: "List tasks through the canonical API execution flow.",
      inputSchema,
      outputSchema,
      isMutating: false,
      execute: async (input, ctx) => {
        const cases = ctx.cases as ExpectedCasesMap | undefined;
        const result = await cases?.tasks?.task_list?.api?.handler(input);

        if (!result?.success || !result.data) {
          throw toAppCaseError(result?.error, "task_list API failed");
        }

        return result.data;
      },
    };
  }

  public mcp(): AgenticMcpContract {
    return {
      enabled: true,
      name: "task_list",
      title: "List Tasks",
      description:
        "Inspect the current task board state through the canonical APP task_list API flow.",
      metadata: {
        category: "tasks",
        mutating: false,
      },
    };
  }

  public rag(): AgenticRagContract {
    return {
      topics: ["task_management", "board_state", "task_grounding"],
      resources: [
        {
          kind: "case",
          ref: "tasks/task_list",
          description: "Canonical board-state capability used for grounding agent decisions.",
        },
        {
          kind: "case",
          ref: "tasks/task_move",
          description: "Related board mutation capability that depends on accurate task identification.",
        },
        {
          kind: "case",
          ref: "tasks/task_create",
          description: "Related board mutation capability that adds new work into the backlog.",
        },
      ],
      hints: [
        "Use task_list before ambiguous mutations so the agent grounds itself on persisted ids and statuses.",
      ],
      scope: "project",
      mode: "recommended",
    };
  }

  public policy(): AgenticPolicy {
    return {
      requireConfirmation: false,
      riskLevel: "low",
      executionMode: "direct-execution",
    };
  }

  public examples(): AgenticExample<TaskListInput, TaskListOutput>[] {
    return super.examples();
  }

  public async test(): Promise<void> {
    this.validateDefinition();

    const definition = this.definition();
    if (definition.tool.isMutating) {
      throw new Error("test: task_list agentic must be read-only");
    }

    if (!definition.rag?.resources?.length) {
      throw new Error("test: task_list should publish semantic RAG resources");
    }

    const example = this.examples().find((item) => item.name === "board_with_cards");
    if (!example) {
      throw new Error("test: task_list example must exist");
    }

    const mockCtx: AgenticContext = {
      correlationId: "task-list-agentic-test",
      logger: this.ctx.logger,
      cases: {
        tasks: {
          task_list: {
            api: {
              handler: async () => ({
                success: true,
                data: example.output,
              }),
            },
          },
        },
      },
    };

    const result = await this.tool().execute(example.input, mockCtx);
    if (result.tasks.length !== example.output.tasks.length) {
      throw new Error("test: task_list tool must return the mocked task collection");
    }

    let propagatedError: unknown;
    try {
      await this.tool().execute(example.input, {
        correlationId: "task-list-agentic-failure-test",
        logger: this.ctx.logger,
        cases: {
          tasks: {
            task_list: {
              api: {
                handler: async () => ({
                  success: false,
                  error: {
                    code: "INTERNAL",
                    message: "Persisted task data is invalid",
                  },
                }),
              },
            },
          },
        },
      });
    } catch (error: unknown) {
      propagatedError = error;
    }

    if (!(propagatedError instanceof AppCaseError)) {
      throw new Error("test: task_list must propagate AppCaseError failures");
    }

    if (propagatedError.code !== "INTERNAL") {
      throw new Error("test: task_list must preserve API error codes");
    }
  }
}

File: examples/flutter/cases/tasks/task_list/task_list.agentic.case.dart
import '../../../core/agentic.case.dart';
import '../../../core/api.case.dart';
import '../../../core/shared/app_structural_contracts.dart';
import 'task_list.domain.case.dart';

typedef _TaskListHandler =
    Future<ApiResponse<TaskListOutput>> Function(TaskListInput input);

class TaskListAgentic extends BaseAgenticCase<TaskListInput, TaskListOutput> {
  TaskListAgentic(super.ctx);

  @override
  TaskListDomain domain() {
    return TaskListDomain();
  }

  @override
  AgenticDiscovery discovery() {
    return AgenticDiscovery(
      name: domainCaseName() ?? 'task_list',
      description:
          domainDescription() ??
          'List persisted task cards for board rendering.',
      category: 'tasks',
      tags: const <String>['tasks', 'listing', 'board'],
      aliases: const <String>[
        'list_board_tasks',
        'show_board',
        'inspect_board_state',
      ],
      capabilities: const <String>['task_listing', 'board_grounding'],
      intents: const <String>[
        'list tasks',
        'show the board',
        'show current tasks',
      ],
    );
  }

  @override
  AgenticExecutionContext context() {
    return const AgenticExecutionContext(
      requiresAuth: false,
      dependencies: <String>['task_list.domain', 'task_list.api'],
      preconditions: <String>['The persisted task store must be readable.'],
      constraints: <String>[
        'This capability is read-only.',
        'No filters or pagination are supported in v1.',
      ],
      notes: <String>[
        'Use this capability to ground follow-up decisions before mutating the board.',
      ],
    );
  }

  @override
  AgenticPrompt prompt() {
    return const AgenticPrompt(
      purpose:
          'List all persisted tasks so an agent can inspect the board state.',
      whenToUse: <String>[
        'When the user asks to see the board or current tasks.',
        'Before moving a task when the user has not provided an exact task identifier.',
      ],
      whenNotToUse: <String>[
        'When the user wants to create a new task.',
        'When the user already provided a precise task id for a direct move operation.',
      ],
      constraints: <String>[
        'Do not claim support for filters or search in this v1 example.',
      ],
      reasoningHints: <String>[
        'Treat task_list as the canonical grounding step before ambiguous task mutations.',
      ],
      expectedOutcome:
          'A flat array of task objects ordered by createdAt descending.',
    );
  }

  @override
  AgenticToolContract<TaskListInput, TaskListOutput> tool() {
    final inputSchema = domainInputSchema();
    final outputSchema = domainOutputSchema();

    if (inputSchema == null || outputSchema == null) {
      throw Exception('task_list.agentic requires domain schemas');
    }

    return AgenticToolContract<TaskListInput, TaskListOutput>(
      name: 'task_list',
      description: 'List tasks through the canonical API execution flow.',
      inputSchema: inputSchema,
      outputSchema: outputSchema,
      parseInput: (dynamic value) =>
          value is TaskListInput ? value : TaskListInput.fromJson(value),
      isMutating: false,
      execute: (TaskListInput input, AgenticContext runtimeContext) async {
        final handler = _resolveHandler(runtimeContext);
        final result = await handler(input);

        if (!result.success || result.data == null) {
          throw toAppCaseError(result.error, 'task_list API failed');
        }

        return result.data!;
      },
    );
  }

  @override
  AgenticMcpContract mcp() {
    return const AgenticMcpContract(
      enabled: true,
      name: 'task_list',
      title: 'List Tasks',
      description:
          'Inspect the current task board state through the canonical APP task_list API flow.',
      metadata: <String, dynamic>{'category': 'tasks', 'mutating': false},
    );
  }

  @override
  AgenticRagContract rag() {
    return const AgenticRagContract(
      topics: <String>['task_management', 'board_state', 'task_grounding'],
      resources: <RagResource>[
        RagResource(
          kind: 'case',
          ref: 'tasks/task_list',
          description:
              'Canonical board-state capability used for grounding agent decisions.',
        ),
        RagResource(
          kind: 'case',
          ref: 'tasks/task_move',
          description:
              'Related board mutation capability that depends on accurate task identification.',
        ),
        RagResource(
          kind: 'case',
          ref: 'tasks/task_create',
          description:
              'Related board mutation capability that adds new work into the backlog.',
        ),
      ],
      hints: <String>[
        'Use task_list before ambiguous mutations so the agent grounds itself on persisted ids and statuses.',
      ],
      scope: 'project',
      mode: 'recommended',
    );
  }

  @override
  AgenticPolicy policy() {
    return const AgenticPolicy(
      requireConfirmation: false,
      riskLevel: 'low',
      executionMode: 'direct-execution',
    );
  }

  _TaskListHandler _resolveHandler(AgenticContext runtimeContext) {
    final tasksDomain = runtimeContext.cases?['tasks'];
    if (tasksDomain is! Map<String, dynamic>) {
      throw const AppCaseError(
        'INTERNAL',
        'task_list requires ctx.cases.tasks',
      );
    }

    final taskList = tasksDomain['task_list'];
    if (taskList is! Map<String, dynamic>) {
      throw const AppCaseError(
        'INTERNAL',
        'task_list requires ctx.cases.tasks.task_list',
      );
    }

    final handler = taskList['apiHandler'];
    if (handler is _TaskListHandler) {
      return handler;
    }

    if (handler is Function) {
      return (TaskListInput input) async {
        final response = await Function.apply(handler, <dynamic>[input]);
        if (response is ApiResponse) {
          return ApiResponse<TaskListOutput>(
            success: response.success,
            data: response.data as TaskListOutput?,
            error: response.error,
            statusCode: response.statusCode,
          );
        }

        throw const AppCaseError(
          'INTERNAL',
          'task_list canonical API handler returned an invalid response',
        );
      };
    }

    throw const AppCaseError(
      'INTERNAL',
      'task_list requires a canonical API handler',
    );
  }

  @override
  Future<void> test() async {
    validateDefinition();

    final definition = this.definition();
    if (definition.tool.isMutating == true) {
      throw Exception('test: task_list agentic must be read-only');
    }

    if ((definition.rag?.resources?.isEmpty ?? true)) {
      throw Exception('test: task_list should publish semantic RAG resources');
    }

    final example = examples().firstWhere(
      (AgenticExample<TaskListInput, TaskListOutput> item) =>
          item.name == 'board_with_cards',
    );
    Future<ApiResponse<TaskListOutput>> successHandler(TaskListInput _) async {
      return ApiResponse.success(example.output);
    }

    Future<ApiResponse<TaskListOutput>> failureHandler(TaskListInput _) async {
      return ApiResponse.failure(
        const AppErrorData(
          code: 'INTERNAL',
          message: 'Persisted task data is invalid',
        ),
      );
    }

    final result = await tool().execute(
      example.input,
      AgenticContext(
        correlationId: 'task-list-agentic-test',
        logger: ctx.logger,
        cases: <String, dynamic>{
          'tasks': <String, dynamic>{
            'task_list': <String, dynamic>{'apiHandler': successHandler},
          },
        },
      ),
    );

    if (result.tasks.length != example.output.tasks.length) {
      throw Exception(
        'test: task_list tool must return the mocked task collection',
      );
    }

    Object? propagatedError;
    try {
      await tool().execute(
        example.input,
        AgenticContext(
          correlationId: 'task-list-agentic-failure-test',
          logger: ctx.logger,
          cases: <String, dynamic>{
            'tasks': <String, dynamic>{
              'task_list': <String, dynamic>{'apiHandler': failureHandler},
            },
          },
        ),
      );
    } catch (error) {
      propagatedError = error;
    }

    if (propagatedError is! AppCaseError) {
      throw Exception('test: task_list must propagate AppCaseError failures');
    }

    if (propagatedError.code != 'INTERNAL') {
      throw Exception('test: task_list must preserve API error codes');
    }
  }
}
